api/auth/external_auth.proto

```diff
@@ -112,6 +103,9 @@ message AttributeContext {
     // The network protocol used with the request, such as
     // "http/1.1", "spdy/3", "h2", "h2c"
     string protocol = 10;
+
+    // The timestamp when the proxy receives the first byte of the request.
+    google.protobuf.Timestamp time = 11;
```

Comment:
Why not have a common request object that is embedded in each concrete type? (Again please see how we are doing access logging which is basically the same idea). Otherwise we will end up with a bunch of duplicated fields. E.g., I think people are going to almost immediately ask for incoming TLS information which is identical to what we are doing in access log. Can that be shared?

Reply:
@mattklein123 @mandarjog I suggest we have a video conference on Monday, so we can discuss the trade offs. I will explain a few points in writing for other people to see.

Attributes are the fundamental building block for Istio and Istio components, including Envoy. Different components exchange attributes for telemetry (e.g. generating logs, metrics, events, traces) and policies (e.g. allowing request.path ~= /v1/). It is the common language between components and Istio users.

We need the attributes to be simple, stable, and easy-to-use, because millions of developers and operators will use them, e.g. writing policies. The number of Envoy and Istio developers will be much smaller. So it is more important to favor external people. FWIW, when Google design their public protos, https://github.com/googleapis/googleapis, the consistency with internal ones was not a consideration.

For Istio, attributes are below of logs, metrics, traces in term of layering. So it is fine for logs to use attributes, but attributes cannot use logs. Because more people will use attributes, so the usability of attributes should be much better than logs.

Because Envoy and Istio were developed in parallel, we didn't converge Envoy and Istio attributes at the start. This leads to the current situation. For Istio, the attribute `request.path` is well established and widely used, and people like it. Renaming it to request.http_request.path or http_request.path would be a regression on usability and won't be accepted. If we use different schemas between Envoy and Istio, it leads to different naming and a runtime translation for every request, e.g. both human cost and runtime cost.

We have a few options below:

```
message Request {
  oneof {
    http_request;
    grpc_request;
    tcp_request;
  }
}
```
This option creates different namespace between Envoy and Istio by extra level nesting. There is also massive duplication between http and grpc definitions.

```
message AttributeContext {
  HTTPRequest request;
  grpc_request;
  tcp_request;
}
```
In this model, Envoy and Istio share the same namespace for HTTP request, which is 99% use cases, and grpc and tcp will add new attributes, but grpc doesn't need to duplicate http stuff, which is a big saving.

```
message AttributeContext {
  HTTPRequest request;
  TCPRequest tpc;
  TLSRequest tls;
}

message HttpRequest {
  // Normal http attributes.
  Auth auth;
  GRPCRequest grpc;
}
```

This option would be my preference. All HTTP based protocols share the outer HTTPRequest. This allows attribute sharing, and shares the same namespace as Istio. Non-HTTP protocols, such as TPC, mTLS, will be added to the top level.

In general, HTTP runs on mTLS on TPC on IP, so the current `oneof` design does not match the reality. This option would solve both usability and accuracy problems.

Currently, Istio uses `connection` for tcp attribute namespace. We can rename Istio attributes to `tcp` namespace to match Envoy. @mandarjog and I will try very hard to align the attribute definition across Envoy, Istio, and other communities. We think this is the best for our users.